README
			      Mortgage Calculator

What is this?
-------------

Mortgage calculator (mcalc for short) is an web application which can
calculate both monthly annuity payment and the amount you can get for a given
mortgage. It also can show amortization schedule.

The latest version
------------------

The latest version can always be obtained from the Georgi Sotirov's projects
page at http://sotirov-bg.net/~gsotirov/projects/

Installation
-------------

Please, see the file called INSTALL for details on the installation.

Copying
--------

Please, see the file called COPYING.



ChangeLog
Mortgage Calculator Changes Log

Note: Bugs could be found at https://sotirov-bg.net/bugzilla/

mcalc 0.5.1 (2019-02-17)
------------------------------------------------------------------------------
* Fix: Update speech style sheet for conformance with CSS 3 specification and
       so it validates without errors and warnings
* Fix: Move tab order on the buttons (was on the image instead)

mcalc 0.5.0 (2015-08-27)
------------------------------------------------------------------------------
This is major new release, including many new features and several bug fixes.
* New: Possibility for calculation with one-time, yearly and monthly
       taxes (see bug 144)
* New: Possibility for calculation with several interest rates for different
       periods (see bugs 145 and 153)
* New: Option to calculate periodical taxes on the last installment of
       the period (see bug 151)
* New: Display total amount of interest to be paid (see bug 82 and 152)
* New: Alternatively enable/disable inputs for amount or payment (see bug 162)
* New: Improved print layout (see bug 155)
* New: HTML source is now in HTML5 (see bug 161)
* New: The calculator now has icon (see bug 163)
* Fix: Results of calculation were not cleared on reset (see bug 147)
* Fix: JavaScript issues in Chrome with default function argument and script
       defer on loading (see bug 154)
* Fix: Amortization table appeared mangled in InternetExplorer (see bug 156)
* Fix: Move DOCTYPE and charset definitions in the beginning to avoid quirks
       mode of InternetExplorer (see bug 159)

mcalc 0.2.5 (2011-03-15)
------------------------------------------------------------------------------
This release includes just one fix found by Ivan Mednev.
* Fix a wrong calculation with numbers higher than "999 999.99" (see bug 126)

mcalc 0.2.4 (2006-12-07)
------------------------------------------------------------------------------
* Fixed program and version information on all files

mcalc 0.2.3 (2006-12-06)
------------------------------------------------------------------------------
* The 'table' is now referred as 'amortization plan'
* The calculate button can now be activated both with the mouse and
  the keyboard
* Fixed HTML syntax warnings

mcalc 0.2.2 (2006-03-03)
------------------------------------------------------------------------------
* Now the final amount will always be zero (see bug 20)
* Respectively the return amount calculation is updated (see bug 21)

mcalc 0.2.1 (2006-02-16)
------------------------------------------------------------------------------
* Initial public release



mcc.js
/**
 * Mortgage Calculator
 * Copyright (C) 2004-2021  Georgi D. Sotirov
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * ---------------------------------------------------------------------------
 * @file Mortgage Calculator Core JavaScript
 * @version 0.5.2
 * @author Georgi D. Sotirov <gdsotirov@gmail.com>
 */

/**
 * Calculates the period payment for annuity mortgages
 * @param {number} interest The yearly interest for the credit
 * @param {number} amount The amount for the credit
 * @param {number} periods The periods for the credit
 * @returns Amount to pay for the period
 */
function calc_period_payment(interest, amount, periods) {
  var payment;

  if ( interest > 0.0 ) {
    var period_interest = interest / 100 / 12;
    payment = round((period_interest * amount) / (1 - Math.pow(1 + period_interest, -periods)), 2);
  }
  else {
    payment = round(amount / periods, 2);
  }

  return payment;
}

/**
 * Calculates the total amount that can be given
 * @param {number} interest The yearly interest for the credit
 * @param {number} payment The period payment
 * @param {number} periods The periods for the credit
 * @returns Total amount that can be given
 */
function calc_total_amount(interest, payment, periods) {
  var amount;

  if ( interest > 0.0 ) {
    var period_interest = interest / 100 / 12;
    amount = (payment * (1 - Math.pow(1 + period_interest, -periods))) / period_interest;
  }
  else {
    amount = payment * periods;
  }

  return amount;
}

/**
 * Builds mortgage amortization plan
 * @param {object} credit Object with all details of the credit
 * @returns An object with amortization table and total amounts
 */
function calc_plan(credit) {
  var balance = credit.amount;
  var annual_tax    = 0.0;
  var monthly_tax   = 0.0;
  var onetime_tax   = 0.0;
  var total_ret_amt = 0.0;
  var total_ints    = 0.0;
  var total_taxes   = 0.0;
  var Rows = new Array();

  /* calculate one time tax */
  if ( credit.onetime_tax_rate || credit.onetime_tax_amt )
  {
    if ( credit.onetime_tax_rate )
      onetime_tax += round(balance * (credit.onetime_tax_rate / 100), 2);
    if ( credit.onetime_tax_amt )
      onetime_tax += round(credit.onetime_tax_amt, 2);

    Rows[0] = new Array(0,
                        balance,
                        0.0,
                        0.0,
                        0.0,
                        0.0,
                        balance,
                        credit.onetime_tax_rate,
                        onetime_tax);

    total_taxes += onetime_tax;
    total_ret_amt += onetime_tax;
  }

  for ( var i = 1; i <= credit.periods; ++i ) {
    var interest = 0.0;
    var interest_rate = 0.0;
    var taxes_rate = 0.0;

    if ( credit.int_periods && i <= credit.int_periods )
    {
      interest = round(balance * (credit.int_rate / 100 / 12), 2);
      interest_rate = credit.int_rate
    }
    else
    if ( credit.int_periods2 && i <= (credit.int_periods + credit.int_periods2) )
    {
      /* Recalculate for the remaining amount and term */
      if ( (i - credit.int_periods) == 1 )
        credit.payment = calc_period_payment(credit.int_rate2, balance, credit.periods - credit.int_periods);
      interest = round(balance * (credit.int_rate2 / 100 / 12), 2);
      interest_rate = credit.int_rate2;
    }
    else
    {
      /* Recalculate for the remaining amount and term */
      if ( (i - (credit.int_periods + credit.int_periods2)) == 1 )
        credit.payment = calc_period_payment(credit.int_rate3, balance, credit.periods - (credit.int_periods + credit.int_periods2));
      interest = round(balance * (credit.int_rate3 / 100 / 12), 2);
      interest_rate = credit.int_rate3;
    }
    total_ints += interest;

    /* Monthly tax on each installment */
    if ( credit.monthly_tax_rate || credit.monthly_tax_amt )
    {
      if (credit.monthly_tax_rate )
        monthly_tax += round(balance * (credit.monthly_tax_rate / 100), 2);
        taxes_rate += credit.monthly_tax_rate;
      if (credit.monthly_tax_amt )
        monthly_tax += round(credit.monthly_tax_amt, 2);

      total_taxes += monthly_tax;
    }

    if ( credit.periods == i ) { // last installment
      var last_payment = balance + interest;
      Rows[i] = new Array(i,
                          balance,
                          interest_rate,
                          interest,
                          balance,
                          last_payment,
                          0.0,
                          taxes_rate,
                          monthly_tax);
      total_ret_amt += last_payment + monthly_tax;
    }
    else {
      var capital     = credit.payment - interest;
      var new_balance = round(balance + interest - credit.payment, 2);
      /* Annual tax on last/first installment of every year since the beginning */
      /* e.g. every December or January depending on selected type */
      if ( (i != 1 && i % 12 == credit.annual_tax_type)
           && (credit.annual_tax_rate || credit.annual_tax_amt) )
      {
        if ( credit.annual_tax_rate )
          annual_tax += round(balance * (credit.annual_tax_rate / 100), 2);
          taxes_rate += credit.annual_tax_rate;
        if ( credit.annual_tax_amt )
          annual_tax += round(credit.annual_tax_amt, 2);

        total_taxes += annual_tax;
      }

      Rows[i] = new Array(i,
                          balance,
                          interest_rate,
                          interest,
                          capital,
                          credit.payment,
                          new_balance,
                          taxes_rate,
                          annual_tax + monthly_tax);
      total_ret_amt += credit.payment + annual_tax + monthly_tax;
    }
    balance = new_balance;
    monthly_tax = 0.0;
    annual_tax = 0.0;
  }

  return {am_tbl : Rows,
          tot_int: total_ints,
          tot_tax: total_taxes,
          tot_ret: total_ret_amt
         };
}

/**
 * Rounds a number to given decimal places
 * @param {number} number A number
 * @param {number} places Decimal places
 * @returns Rounded number
 */
function round(number, places) {
  return Math.round(number * Math.pow(10, places)) / Math.pow(10, places);
}


sprintf.js
/* Function printf(format_string,arguments...)
 * Javascript emulation of the C printf function (modifiers and argument types
 *    "p" and "n" are not supported due to language restrictions)
 *
 * Copyright 2003 K&L Productions. All rights reserved
 * http://www.klproductions.com
 *
 * Terms of use: This function can be used free of charge IF this header is not
 *               modified and remains with the function code.
 *
 * Legal: Use this code at your own risk. K&L Productions assumes NO resposibility
 *        for anything.
 ********************************************************************************/

/* Modified by Georgi D. Sotirov <gdsotirov@dir.bg>
 *
 * 2005-08-15 Added support for %% which should escape % character
 */

function sprintf(fstring) {
  var pad = function(str, ch, len) {
    var ps = '';
    for ( var i = 0; i < Math.abs(len); i++ )
      ps += ch;
    return len > 0 ? str + ps : ps + str;
  }

  var processFlags = function(flags, width, rs, arg) {
    var pn = function(flags, arg, rs) {
      if ( arg >= 0) {
        if ( flags.indexOf(' ') >= 0 )
          rs = ' ' + rs;
        else if ( flags.indexOf('+') >= 0 )
          rs = '+' + rs;
      }
      else
        rs = '-' + rs;
      return rs;
    }

    var iWidth = parseInt(width, 10);

    if ( width.charAt(0) == '0' ) {
      var ec = 0;
      if ( flags.indexOf(' ') >= 0 || flags.indexOf('+') >= 0 )
        ec++;
      if ( rs.length < (iWidth - ec) )
        rs = pad(rs, '0', rs.length - (iWidth - ec));
      return pn(flags, arg, rs);
    }

    rs = pn(flags, arg, rs);

    if ( rs.length < iWidth ) {
      if ( flags.indexOf('-') < 0 )
        rs = pad(rs, ' ', rs.length - iWidth);
      else
        rs = pad(rs, ' ', iWidth - rs.length);
    }
    return rs;
  }

  var converters = new Array();

  converters['c'] = function(flags, width, precision, arg) {
    if ( typeof(arg) == 'number' )
      return String.fromCharCode(arg);
    if ( typeof(arg) == 'string' )
      return arg.charAt(0);
    return '';
  }

  converters['d'] = function(flags, width, precision, arg) {
    return converters['i'](flags, width, precision, arg);
  }

  converters['u'] = function(flags, width, precision, arg) {
    return converters['i'](flags, width, precision, Math.abs(arg));
  }

  converters['i'] = function(flags, width, precision, arg) {
    var iPrecision = parseInt(precision);
    var rs = ((Math.abs(arg)).toString().split('.'))[0];
    if ( rs.length < iPrecision )
      rs = pad(rs, ' ', iPrecision - rs.length);
    return processFlags(flags, width, rs, arg);
  }

  converters['E'] = function(flags, width, precision, arg) {
    return (converters['e'](flags, width, precision, arg)).toUpperCase();
  }

  converters['e'] =  function(flags, width, precision, arg) {
    iPrecision = parseInt(precision);
    if ( isNaN(iPrecision) )
      iPrecision = 6;
    rs = (Math.abs(arg)).toExponential(iPrecision);
    if ( rs.indexOf('.') < 0 && flags.indexOf('#') >= 0 )
      rs = rs.replace(/^(.*)(e.*)$/,'$1.$2');
    return processFlags(flags,width,rs,arg);
  }

  converters['f'] = function(flags, width, precision, arg) {
    iPrecision = parseInt(precision);
    if ( isNaN(iPrecision) )
      iPrecision = 6;
    rs = (Math.abs(arg)).toFixed(iPrecision);
    if ( rs.indexOf('.') < 0 && flags.indexOf('#') >= 0 )
      rs = rs + '.';
    return processFlags(flags, width, rs, arg);
  }

  converters['G'] = function(flags, width, precision, arg) {
    return (converters['g'](flags, width, precision, arg)).toUpperCase();
  }

  converters['g'] = function(flags, width, precision, arg) {
    iPrecision = parseInt(precision);
    absArg = Math.abs(arg);
    rse = absArg.toExponential();
    rsf = absArg.toFixed(6);
    if ( !isNaN(iPrecision) ) {
      rsep = absArg.toExponential(iPrecision);
      rse = rsep.length < rse.length ? rsep : rse;
      rsfp = absArg.toFixed(iPrecision);
      rsf = rsfp.length < rsf.length ? rsfp : rsf;
    }
    if ( rse.indexOf('.') < 0 && flags.indexOf('#') >= 0 )
      rse = rse.replace(/^(.*)(e.*)$/,'$1.$2');
    if ( rsf.indexOf('.') < 0 && flags.indexOf('#') >= 0 )
      rsf = rsf + '.';
    rs = rse.length < rsf.length ? rse : rsf;
    return processFlags(flags, width, rs, arg);
  }

  converters['o'] = function(flags, width, precision, arg) {
    var iPrecision = parseInt(precision);
    var rs = Math.round(Math.abs(arg)).toString(8);
    if ( rs.length < iPrecision)
      rs = pad(rs, ' ', iPrecision - rs.length);
    if ( flags.indexOf('#') >= 0 )
      rs = '0' + rs;
    return processFlags(flags,width,rs,arg); 
  }

  converters['X'] = function(flags, width, precision, arg) {
    return (converters['x'](flags, width, precision, arg)).toUpperCase();
  }

  converters['x'] = function(flags, width, precision, arg) {
    var iPrecision = parseInt(precision);
    arg = Math.abs(arg);
    var rs = Math.round(arg).toString(16);
    if ( rs.length < iPrecision)
      rs = pad(rs, ' ', iPrecision - rs.length);
    if ( flags.indexOf('#') >= 0 )
      rs = '0x' + rs;
    return processFlags(flags,width,rs,arg); 
  }

  converters['s'] = function(flags, width, precision, arg) {
    var iPrecision = parseInt(precision);
    var rs = arg;
    if ( rs.length > iPrecision)
      rs = rs.substring(0, iPrecision);
    return processFlags(flags,width,rs,0);
  }

  converters['%'] = function(flags, width, precision, arg) {
    return '%';
  }

  var fpRE = /([^%]*)%([-+ #]*)(\d*)\.?(\d*)([cdieEfFgGosuxX%])([^%]*)/g;
  var retstr = "";
  var index = 1;

  while ( fps = fpRE.exec(fstring) ) {
    if ( arguments[index] || fps[5] == '%' )
      retstr += converters[fps[5]](fps[2], fps[3], fps[4], arguments[index]);
    retstr = fps[1] + retstr + fps[6];
    if ( fps[5] != '%' )
      ++index;
  }
  if (index != 1)
    return retstr;
  else
    return fstring;
}



mcalc.js
/**
 * Mortgage Calculator
 * Copyright (C) 2004-2021  Georgi D. Sotirov
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * ---------------------------------------------------------------------------
 * @file Mortgage Calculator UI JavaScript
 * @version 0.5.2
 * @author Georgi D. Sotirov <gdsotirov@gmail.com>
 */

var uisPlsFillAmount = 0;
var uisPlsCorrAmount = 1;
var uisPlsFillPayment = 2
var uisPlsCorrPayment = 3;
var uisPlsChsePeriod = 4;
var uisPlsFillInterest = 5;
var uisPlsCorrInterest = 6;
var uisDate = 7;
var uisBalance = 8;
var uisIntRate = 9;
var uisInterest = 10;
var uisCapital = 11;
var uisPayment = 12;
var uisOutstanding = 13;
var uisTaxesRate = 14;
var uisTaxes   = 15;

var UIStringsBG = new Array(
/*  0 */ "Моля, попълнете полето Сума!",
/*  1 */ "Моля, задайте правилна стойност в полето Сума!\nНапример: 10000, 15500, 20100.55",
/*  2 */ "Моля, попълнете полето Вноска!",
/*  3 */ "Моля, задайте правилна стойност в полето Вноска!\nНапример: 500, 750, 1200.5",
/*  4 */ "Моля, изберете срок на кредита в години и/или месеци!",
/*  5 */ "Моля, попълнете полетата Лихвен процент!",
/*  6 */ "Моля, задайте правилна стойност в полето Годишен лихвен процент!\nНапример: 10.5, 12.75, 11",
/*  7 */ "Месец",
/*  8 */ "Салдо главница",
/*  9 */ "Лихвен %",
/* 10 */ "Вноска лихва",
/* 11 */ "Вноска главница",
/* 12 */ "Вноска общо",
/* 13 */ "Оставащо",
/* 14 */ "Такси %",
/* 15 */ "Такси"
);

var UIStringsEN = new Array(
/*  0 */ "Please, fill in the Amount field!",
/*  1 */ "Please, fill in correct value in the Amount field! Example: 10000, 15500, 20100.55",
/*  2 */ "Please, fill in the Payment field!",
/*  3 */ "Please, fill in correct value in the Payment field! Example: 500, 750, 1200.5",
/*  4 */ "Please, choose the credit term in years and/or months!",
/*  5 */ "Please, fill in the Interest rate fields!",
/*  6 */ "Please, fill in correct value in the Interest field! Example: 10000, 15500, 20100.55",
/*  7 */ "Month",
/*  8 */ "Capital Balance",
/*  9 */ "Interest rate",
/* 10 */ "Interest payment",
/* 11 */ "Capital payment",
/* 12 */ "Payment total",
/* 13 */ "Outstanding",
/* 14 */ "Taxes rate",
/* 15 */ "Taxes"
);

/**
 * Loads UI string base on language setting of html tag
 * @param {number} id UI string identifier
 * @returns String in specific language.
 */
function loadUIString(id) {
  var htmltags = document.getElementsByTagName("html");
  var lang = htmltags[0].lang;

  if ( lang == "en" ) {
    return UIStringsEN[id];
  }
  else if ( lang == "bg" ) {
    return UIStringsBG[id];
  }
  else return "???";
}

/**
 * Formats number with specified precision
 * @param {number} number Number to format
 * @param {number} places Precision in decimal places
 * @returns Formatted string representation of the number
 */
function formatNumber(number, places = 2) {
  var num = new NumberFormat();

  /* Configure number formatting */
  num.setInputDecimal('.');
  num.setSeparators(true, ' ', '.');
  num.setPlaces(places, false);
  num.setCurrency(false);
  num.setCurrencyPosition(num.LEFT_OUTSIDE);
  num.setNegativeFormat(num.LEFT_DASH);
  num.setNegativeRed(false);
  num.setNumber(number);

  return num.toFormatted();
}

/**
 * Formats the value of and HTML input element
 * @param {object} obj An HTML input element
 * @param {number} places Precision in decimal places
 */
function formatField(obj, places = 2) {
  if ( obj.value != 0 )
  {
    obj.value = formatNumber(obj.value.replace(",", ".").replace(/\s+/g, ""), places);
  }
  else
  {
    obj.value = "";
  }
}

/**
 * Checks the value of an HTML input element
 * @param {object} fld An HTML input element
 * @param {string} type Value type. Either 'float' or 'int'
 * @param {number} uisFill UI string index for when field value is empty
 * @param {number} uisCorr UI string index for when field value is wrong
 * @returns True if the value in the field is correct, otherwise false
 */
function checkField(fld, type, uisFill, uisCorr) {
  var val;

  if ( fld.value == "" ) {
    alert(loadUIString(uisFill));
    fld.focus();
    return false;
  }

  if ( type != "string" ) {
    if ( type == "float" )
      val = parseFloat(fld.value);
    else if ( type == "int" )
      val = parseInt(fld.value);
    if ( isNaN(val) ) {
      alert(loadUIString(uisCorr));
      fld.focus();
      return false;
    }
  }

  return true;
}

/**
 * Removes all children elements of an element
 * @param {object} elem An HTML element (any)
 */
 function removeAllChildren(elem) {
  while ( elem.firstChild ) {
    elem.removeChild(elem.lastChild);
  }
}

/**
 * Retrieves the value of a radio button
 * @param {object} radio An HTML radio buttons input
 * @returns The selected radio button value, otherwise undefined
 */
function getRadioValue(radio) {
  var i = 0;
  while ( i < radio.length ) {
    if ( radio[i].checked )
      return radio[i].value;
    ++i;
  }
  return;
}

/**
 * Parses floating value by first removing spaces and replacing , with .
 * @param {string} str String with float value
 * @returns Float value
 */
function getFloatValue(str) {
  return parseFloat(str.replace(/\s+/g, "").replace(",", "."));
}

/**
 * Hides select for months if maximum term of 30 years is chosen
 */
function lockMonths() {
  var year = parseInt(document.forms.CalcForm.TermY.value);
  if ( !isNaN(year) ) {
    document.forms.CalcForm.TermM.disabled = (year == 30);
    var msel = document.getElementById("MonthSelect");
    if ( year == 30 )
      msel.style.display = "none";
    else
      msel.style.display = "inline";
  }
}

/**
 * Disables either input for Amount or Payment based on the value of Type
 * radio button, because these are mutually exclusive calculation options
 */
function lockType() {
  var type_val = getRadioValue(document.forms.CalcForm.Type);
  document.getElementById("Payment").disabled = (type_val == "payment");
  document.getElementById("Amount" ).disabled = (type_val == "amount" );
}

/**
 * Finds next element in tab index
 * @see https://stackoverflow.com/a/29883167
 * @param {object} el An HTML element
 * @returns The next element in tab index
 */
function findNextTabStop(el) {
  var elmnts = document.querySelectorAll('input, button, select');
  var list = Array.prototype.filter.call(elmnts, function(item) {return item.tabIndex >= "0"});
  var idx = list.indexOf(el);
  return list[idx + 1] || list[0];
}

/**
 * Enables or disables additional interest rates inputs based on what is
 * already entered. User first have to set one interest rate for given period
 * to be able to enter second with second period and finally third for the
 * remaining period.
 * @param {object} elmnt An HTML element
 */
function controlRates(elmnt) {
  var Interest   = document.getElementById("Interest");
  var IntPeriods = document.getElementById("IntPeriods");
  var Interest2  = document.getElementById("Interest2");
  var IntPeriods2= document.getElementById("IntPeriods2");
  var Interest3  = document.getElementById("Interest3");

  IntPeriods.disabled = (Interest.value.length == 0 || Interest.value == "" || Interest.value == 0 );
  Interest2.disabled  = (   (Interest.value.length == 0   || Interest.value == ""   || Interest.value == 0)
                         || (IntPeriods.value.length == 0 || IntPeriods.value == "" || IntPeriods.value == 0)
                        );
  IntPeriods2.disabled= (Interest2.value.length == 0 || Interest2.value == "" || Interest2.value == 0 );
  Interest3.disabled  = (   (Interest2.value.length == 0   || Interest2.value == ""   || Interest2.value == 0)
                         || (IntPeriods2.value.length == 0 || IntPeriods2.value == "" || IntPeriods2.value == 0)
                        );

  /* Focus on next element by tab index */
  var nextElmnt = findNextTabStop(elmnt);
  nextElmnt.focus();
}

/**
 * Shows or hides amortization table
 */
function showPlan() {
  var planc = document.getElementById("PlanContainer");
  if ( document.forms.CalcForm.EnablePlan.checked )
    planc.style.display = "block";
  else
    planc.style.display = "none";
}

/**
 * Checks values in calculator form
 * @returns True if all values are correct, otherwise false
 */
function checkForm() {
  var form = document.forms.CalcForm;
  var type = getRadioValue(form.Type);

  if ( type == "payment" ) {
    if ( !checkField(form.Amount, "float", uisPlsFillAmount, uisPlsCorrAmount) )
      return false;
  }
  else if ( !checkField(form.Payment, "float", uisPlsFillPayment, uisPlsCorrPayment) )
      return false;

  var TermY = parseInt(form.TermY.value);
  var TermM = parseInt(form.TermM.value);
  if ( TermY == 0 && TermM == 0 ) {
    alert(loadUIString(uisPlsChsePeriod));
    form.TermY.focus();
    return false;
  }
  if ( !checkField(form.Interest, "float", uisPlsFillInterest, uisPlsCorrInterest) )
    return false;

  return true;
}

/**
 * Performs calculator form reset
 */
function doReset() {
  removeAllChildren(document.getElementById("TotalInterests"));
  removeAllChildren(document.getElementById("TotalTaxes"));
  removeAllChildren(document.getElementById("TotalReturn"));
  removeAllChildren(document.getElementById("TotalRaise"));
  removeAllChildren(document.getElementById("PlanContainer"));

  document.getElementById("IntPeriods").disabled  = true;
  document.getElementById("Interest2").disabled   = true;
  document.getElementById("IntPeriods2").disabled = true;
  document.getElementById("Interest3").disabled   = true;

  document.forms.CalcForm.TermM.disabled = false;
  document.getElementById("MonthSelect").style.display = "inline";

  document.getElementById("Payment").disabled = true;
  document.getElementById("Amount" ).disabled = false;
}

/**
 * Calculates and displays results and optionally amortization table
 * @returns Always true
 */
function doCalc() {
  var form = document.forms.CalcForm;
  var credit = new Object();

  credit.type = getRadioValue(form.Type);
  credit.amount = getFloatValue(form.Amount.value);
  credit.termY = parseInt(form.TermY.value);
  credit.termM = 0;
  if (credit.termY < 30) credit.termM = parseInt(form.TermM.value);
  credit.periods = credit.termY * 12 + credit.termM;
  credit.int_rate     = getFloatValue(form.Interest.value);
  credit.int_periods  = parseInt(form.IntPeriods.value);
  credit.int_rate2    = getFloatValue(form.Interest2.value);
  credit.int_periods2 = parseInt(form.IntPeriods2.value);
  credit.int_rate3    = getFloatValue(form.Interest3.value);
  if ( !credit.int_periods ) {
    credit.int_periods = credit.periods;
    credit.int_periods2 = 0;
  }
  if ( !credit.int_periods2 && credit.int_rate2 )
  {
    credit.int_periods2 = credit.periods - credit.int_periods;
  }
  credit.payment      = getFloatValue(form.Payment.value);

  credit.annual_tax_rate  = getFloatValue(form.AnnualTaxRate.value);
  credit.annual_tax_amt   = getFloatValue(form.AnnualTaxAmt.value);
  if ( Number(form.AnnualTaxType.value) != 0 ) {
    credit.annual_tax_type  = 1;
  }
  else
  {
    credit.annual_tax_type  = 0;
  }
  credit.monthly_tax_rate = getFloatValue(form.MonthlyTaxRate.value);
  credit.monthly_tax_amt  = getFloatValue(form.MonthlyTaxAmt.value);
  credit.onetime_tax_rate = getFloatValue(form.OneTimeTaxRate.value);
  credit.onetime_tax_amt  = getFloatValue(form.OneTimeTaxAmt.value);

  var enablePlan = form.EnablePlan.checked;

  var Amount  = document.getElementById("Amount");
  var Payment = document.getElementById("Payment");

  if ( credit.type == "payment" ) {
    credit.payment = calc_period_payment(credit.int_rate, credit.amount, credit.periods);
    Payment.value = formatNumber(credit.payment);
  }
  else {
    credit.amount = calc_total_amount(credit.int_rate, credit.payment, credit.periods);
    Amount.value = formatNumber(credit.amount);
  }

  var result = calc_plan(credit);

  var TotalInt    = document.getElementById("TotalInterests");
  var TotalTax    = document.getElementById("TotalTaxes");
  var TotalReturn = document.getElementById("TotalReturn");
  var TotalRaise  = document.getElementById("TotalRaise");
  var AmortPlan   = document.getElementById("PlanContainer");
  removeAllChildren(document.getElementById("TotalInterests"));
  removeAllChildren(document.getElementById("TotalTaxes"));
  removeAllChildren(document.getElementById("TotalReturn"));
  removeAllChildren(document.getElementById("TotalRaise"));
  removeAllChildren(document.getElementById("PlanContainer"));

  var raise = ((result.tot_ret / credit.amount) - 1 ) * 100;

  TotalInt.appendChild(document.createTextNode(formatNumber(result.tot_int)));
  TotalTax.appendChild(document.createTextNode(formatNumber(result.tot_tax)));
  TotalReturn.appendChild(document.createTextNode(formatNumber(result.tot_ret)));
  TotalRaise.appendChild(document.createTextNode(formatNumber(raise, 5) + " %"));

  if ( enablePlan ) {
    var Table = document.createElement("table");
    var Rows  = result.am_tbl;
    Table.setAttribute("class", "tbThinBorder");
    Table.className = "tbThinBorder";
    Table.setAttribute("id", "Table");
    Table.setAttribute("cellspacing", "0");
    makeTableHeader(Table, loadUIString(uisDate),
                           loadUIString(uisBalance),
                           loadUIString(uisIntRate),
                           loadUIString(uisInterest),
                           loadUIString(uisCapital),
                           loadUIString(uisPayment),
                           loadUIString(uisOutstanding),
                           loadUIString(uisTaxesRate),
                           loadUIString(uisTaxes));
    var TableBody = document.createElement("tbody");
    for ( var i = 0; i < Rows.length; ++i ) {
      if ( Rows[i] )
      {
        var Row = Rows[i];
        var elmnt = makeTableRow(TableBody, Row[0],
                                 formatNumber(Row[1]),
                                 formatNumber(Row[2]),
                                 formatNumber(Row[3]),
                                 formatNumber(Row[4]),
                                 formatNumber(Row[5]),
                                 formatNumber(Row[6]),
                                 formatNumber(Row[7]),
                                 formatNumber(Row[8]));

        if ( i % 12 == 0 )
          elmnt.setAttribute("style", "background-color: #eeeeee;");
      }
    }
    Table.appendChild(TableBody);
    AmortPlan.appendChild(Table);
  }

  return true;
}

/**
 * Constructs results table header
 * @param {object} otable An HTML table element
 */
function makeTableHeader(otable) {
  var new_tr = document.createElement("tr");
  for ( var i = 1; i < arguments.length; ++i ) {
    var new_th = document.createElement("th");
    new_th.appendChild(document.createTextNode(arguments[i]));
    new_tr.appendChild(new_th);
  }
  var new_thead = document.createElement("thead");
  new_thead.appendChild(new_tr);
  otable.appendChild(new_thead);
}

/**
 * Constructs results table data row
 * @param {object} otable An HTML table element
 */
function makeTableRow(otable) {
  var new_tr = document.createElement("tr");
  for ( var i = 1; i < arguments.length; ++i ) {
    var new_td = document.createElement("td");
    new_td.appendChild(document.createTextNode(arguments[i]));
    new_tr.appendChild(new_td);
  }
  otable.appendChild(new_tr);
  return new_tr;
}

